bindings/java/src/main/com/apple/foundationdb/FDBTransaction.java

```diff
@@ -220,6 +228,17 @@ public TransactionOptions options() {
 
 	@Override
 	public void setReadVersion(long version) {
+		/*
+		 * Note that this is done outside of the lock, because we don't want to rely on
+		 * the caller code being particularly efficient, and if we get a bad
+		 * implementation of a eventKeeper, we could end up holding the pointerReadLock for an
+		 * arbitrary amount of time; this would be Bad(TM), so we execute this outside
+		 * the lock, so that in the worst case only the caller thread itself can be hurt
+		 * by bad callbacks.
```

Comment:
I think it makes sense to increment this outside of the lock. We are still expecting usages of `eventKeeper` to be thread-safe, though, and for that implementation to avoid gratuitous synchronization between threads, correct? This seems pretty important, because now a number of operations and other code points have to properly manage concurrent access to this data, and we have to be careful that the way this is being done doesn't cause contention between threads.

Do we intend to provide a reference implementation?

Reply:
`EventKeeper` absolutely needs to be thread-safe. I didn't see any specific annotations inside the java codebase to make that clear (for e.g. code analysis tools), but I did add documentation to the `EventKeeper` interface itself.

As for reference implementations, there is `MapTimer`, we could (it wouldn't be hard), I just didn't because I didn't have an immediate need for it and I generally try to avoid adding classes with no obvious usages. I'm happy to add one if you would desire it though.

Reply:
I guess another way to think about this is to ask whether it is easy for someone to implement it well both from a thread-safety and a performance perspective. If an optimal solution is complicated or non-obvious, it may be nice to have a reference implementation that demonstrates how to do it, and it could be one that we use ourselves. If the right solution is the obvious and fairly easy one, then this is less of a concern.

Reply:
I also like the idea of reference implementation. Not a must, but would be nice I think.

Reply:
added `MapEventKeeper`, which provides a simple implementation of EventKeeper using a backing ConcurrentMap. This should provide an adequate reference implementation.